### ISSUE TITLE
Strange behaviour of cloudstor running on a specific d4aws swarm

### ISSUE THREAD

**Issue by hrajchert** [docker/for-aws#94]
### Expected behavior
When moving service instances between different AZ the relocatable volume should always contain the last state.

When deleting a volume using `docker volume rm <volume-name>` and then creating a new one with the same name, the volume should be empty

### Actual behavior
When moving service instances between different AZ the relocatable volume contains one snapshot before the expected.

When deleting a volume using `docker volume rm <volume-name>` and then creating a new one with the same name, the volume contains the last snapshot.

### Information

  - Full output of the diagnostics from "docker-diagnose" ran from one of the instance
```OK hostname=ip-172-31-38-205-ec2-internal session=1503610029-yecBPbvq5l5skqUpcEs4TFNG56kCBOoj
OK hostname=ip-172-31-29-77-ec2-internal session=1503610029-yecBPbvq5l5skqUpcEs4TFNG56kCBOoj
Done requesting diagnostics.
Your diagnostics session ID is 1503610029-yecBPbvq5l5skqUpcEs4TFNG56kCBOoj
Please provide this session ID to the maintainer debugging your issue.
```

### Steps to reproduce the behavior
*Important:* when trying to reproduce this steps on new d4aws swarms, they worked as expected, but when doing this steps on a swarm I created on the 23rd (yesterday) I always have the problem, so there seems to be something corrupted in the swarm but I cannot figure out what.

Steps for reproduce
1) Create a docker stack with 1 manager and 1 worker using this template
https://editions-us-east-1.s3.amazonaws.com/aws/stable/Docker.tmpl

2) Make sure the manager and the worker are in different AZ 
If not terminate the worker until the instance goes into a different AZ

3) Log into the manager and manually create the following volumes

```
docker volume create \
  -d "cloudstor:aws" \
  --opt ebstype=gp2 \
  --opt size=15 \
  --opt iops=180 \
  --opt backing=relocatable \
  ci_ping-1-vol

docker volume create \
  -d "cloudstor:aws" \
  --opt ebstype=gp2 \
  --opt size=15 \
  --opt iops=180 \
  --opt backing=relocatable \
  ci_ping-2-vol
```

If you check in the aws console under EBS volumes you can see in which AZ those volumes where created.
Chances are that at least one of those wont be in the one needed and will have to move.

4) Create this stack named `ci.yml` and wait for it to be fully running

```
version: "3.3"

services:
  ping:
    image: alpine
    command: ping docker.com
    volumes:
      - ping_vol:/mydata
    deploy:
      replicas: 2

volumes:
  ping_vol:
    external:
      name: "{{.Service.Name}}-{{.Task.Slot}}-vol"
```
$ docker stack deploy -c ci.yml ci

execute 
```
$ docker service ls 

ID                  NAME                IMAGE               NODE                            DESIRED STATE       CURRENT STATE                ERROR               PORTS
oudrs6inphg6        ci_ping.1           alpine:latest       ip-172-31-22-59.ec2.internal    Running             Running 2 minutes ago                            
w1kprholjum1        ci_ping.2           alpine:latest       ip-172-31-11-213.ec2.internal   Running             Running about a minute ago                  
```
and wait until both replicas are running in different hosts. Chances are that at least one of the volumes need to move

5) Exec an ash session in the manager and touch `/mydata/manager.txt`
~ $ docker exec -it 5eaacefc7e7e ash
/ # touch mydata/manager.txt

6) Exec an ash session in the worker and touch `/mydata/worker-a.txt`
~ $ docker exec -it af3ffc72a1a3 ash
/ # touch mydata/worker-a.txt

7) Terminate the worker and see how the swarm moves the service 

Two things should happen, the autoscaling group should create a new worker host but first the swarm should move the second replica from the worker to the manager, and because they are on different AZ the EBS should be snapshoted before creating the EBS in the manager AZ. 

8) In the manager check the volume attached 
Here is where I see the bug that so far I can only reproduce in one swarm in particular.
On new swarms, the content of `mydata` contains `worker-a.txt` as expected, but on my existing swarm I always have a previous snapshot.

The first time I tried I got an empty volume, and after rescaling the service (which moves the volume to the new worker host) I can see the new service contains worker-a, but not the "latest" changes.

On the problematic swarm I can reproduce this 100% of the times, so for now I'll leave it on until we can investigate further. I would not like to encounter this once we're in production :P 

Thanks!

**Comment by ddebroy** [comment]
@hrajchert can you also please post the `docker volume ls` output from the nodes before and after the scale events in the problematic cluster?

**Comment by hrajchert** [comment]
Sure, thanks for the fast response.

Note: I changed the service and volume name to pingg to avoid the second problem listed.

## Preparation for the test, before doing the "service move"
```
$ docker volume create \
  -d "cloudstor:aws" \
  --opt ebstype=gp2 \
  --opt size=15 \
  --opt iops=180 \
  --opt backing=relocatable \
  ci_pingg-1-vol

$ docker volume create \
  -d "cloudstor:aws" \
  --opt ebstype=gp2 \
  --opt size=15 \
  --opt iops=180 \
  --opt backing=relocatable \
  ci_pingg-2-vol

$ docker stack deploy -c ci.yml ci

$ docker service ps ci_pingg
ID                  NAME                IMAGE               NODE                            DESIRED STATE       CURRENT STATE            ERROR               PORTS
p8yhva9id1o8        ci_pingg.1          alpine:latest       ip-172-31-38-205.ec2.internal   Running             Running 49 seconds ago                       
yjbegvsbzd4c        ci_pingg.2          alpine:latest       ip-172-31-29-77.ec2.internal    Running             Running 8 seconds ago       
```

On the manager
```
$ docker volume ls 
DRIVER              VOLUME NAME
local               6ecdff3eea7e390eb02da83712cc0036f60eb9652e1056692bbc219bbc51265b
cloudstor:aws       alb_acme_vol
local               b91f5eb943068d1bd3897db4322ec37c25208ba49e1e3b74c987917e6c965b06
cloudstor:aws       ci_pingg-1-vol
cloudstor:aws       ci_pingg-2-vol
local               sshkey

$ docker inspect ci_pingg-1-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": {},
        "Mountpoint": "/mnt/ebs/ci_pingg-1-vol",
        "Name": "ci_pingg-1-vol",
        "Options": {
            "backing": "relocatable",
            "ebstype": "gp2",
            "iops": "180",
            "size": "15"
        },
        "Scope": "local"
    }
]

$ docker inspect ci_pingg-2-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": {},
        "Mountpoint": "/mnt/ebs/ci_pingg-2-vol",
        "Name": "ci_pingg-2-vol",
        "Options": {
            "backing": "relocatable",
            "ebstype": "gp2",
            "iops": "180",
            "size": "15"
        },
        "Scope": "local"
    }
]
```

On the worker
```
$ docker volume ls
DRIVER              VOLUME NAME
cloudstor:aws       alb_acme_vol
cloudstor:aws       ci_pingg-1-vol
cloudstor:aws       ci_pingg-2-vol
local               sshkey

$ docker inspect ci_pingg-1-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": null,
        "Mountpoint": "/mnt/ebs/ci_pingg-1-vol",
        "Name": "ci_pingg-1-vol",
        "Options": {},
        "Scope": "local"
    }
]

$ docker inspect ci_pingg-2-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": null,
        "Mountpoint": "/mnt/ebs/ci_pingg-2-vol",
        "Name": "ci_pingg-2-vol",
        "Options": {},
        "Scope": "local"
    }
]
```
On the manager, the 2 local fs correspond to a portainer and registry services, the alb_acme_vol is a EFS cloudstor volume that holds the acme certificates obtained by traefik

## The test
To force the service from the worker to move to the manager I terminate the worker from the aws console

The volumes in the manager stays the same, even though the two services are running in the manager, and the `/mydata` folder doesn't have any data (it should have `worker-aaa.txt`
```
$ docker service ps ci_pingg
ID                  NAME                IMAGE               NODE                            DESIRED STATE       CURRENT STATE            ERROR               PORTS
p8yhva9id1o8        ci_pingg.1          alpine:latest       ip-172-31-38-205.ec2.internal   Running             Running 14 minutes ago                       
j47fz5vn68fr        ci_pingg.2          alpine:latest       ip-172-31-38-205.ec2.internal   Running             Running 2 minutes ago                        
yjbegvsbzd4c         \_ ci_pingg.2      alpine:latest       ip-172-31-29-77.ec2.internal    Shutdown            Running 4 minutes ago              

$ docker volume ls
DRIVER              VOLUME NAME
local               6ecdff3eea7e390eb02da83712cc0036f60eb9652e1056692bbc219bbc51265b
cloudstor:aws       alb_acme_vol
local               b91f5eb943068d1bd3897db4322ec37c25208ba49e1e3b74c987917e6c965b06
cloudstor:aws       ci_pingg-1-vol
cloudstor:aws       ci_pingg-2-vol
local               sshkey

$ docker inspect ci_pingg-1-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": {},
        "Mountpoint": "/mnt/ebs/ci_pingg-1-vol",
        "Name": "ci_pingg-1-vol",
        "Options": {
            "backing": "relocatable",
            "ebstype": "gp2",
            "iops": "180",
            "size": "15"
        },
        "Scope": "local"
    }
]

$ docker inspect ci_pingg-2-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": {},
        "Mountpoint": "/mnt/ebs/ci_pingg-2-vol",
        "Name": "ci_pingg-2-vol",
        "Options": {
            "backing": "relocatable",
            "ebstype": "gp2",
            "iops": "180",
            "size": "15"
        },
        "Scope": "local"
    }
]
```

The new worker doesnt have services as expected and the volumes stay the same.
```
$ docker ps
CONTAINER ID        IMAGE                                COMMAND                  CREATED             STATUS              PORTS                NAMES
2b2a0029d159        docker4x/guide-aws:17.06.0-ce-aws2   "/entry.sh"              6 minutes ago       Up 6 minutes                             guide-aws
ef0dafff49a0        docker4x/shell-aws:17.06.0-ce-aws2   "/entry.sh /usr/sb..."   7 minutes ago       Up 6 minutes        0.0.0.0:22->22/tcp   shell-aws

$ docker volume ls
DRIVER              VOLUME NAME
cloudstor:aws       alb_acme_vol
cloudstor:aws       ci_pingg-1-vol
cloudstor:aws       ci_pingg-2-vol
local               sshkey

$ docker inspect ci_pingg-1-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": null,
        "Mountpoint": "/mnt/ebs/ci_pingg-1-vol",
        "Name": "ci_pingg-1-vol",
        "Options": {},
        "Scope": "local"
    }
]

$ docker inspect ci_pingg-2-vol
[
    {
        "Driver": "cloudstor:aws",
        "Labels": null,
        "Mountpoint": "/mnt/ebs/ci_pingg-2-vol",
        "Name": "ci_pingg-2-vol",
        "Options": {},
        "Scope": "local"
    }
]
```

At this moment both services are running on the master, if I scale down and then up again it will move from the manager to the worker (which are still in different AZ). Let me know if you want to test something else while moving.

**Comment by hrajchert** [comment]
Hi @ddebroy! sorry to bother you, but is there any other test that you might want to do so I can delete the swarm and release the resources?

Thanks!

**Comment by ddebroy** [comment]
Hi @hrajchert please go ahead and delete/release the resources. Let's keep this bug open for tracking the issue and update here with any relevant findings.

**Comment by ddebroy** [comment]
@hrajchert an update on this: I managed to get a repro of the issue you reported. The repro I got was using two nodes within the same AZ (i.e. with no EBS snapshots involved) - just an attach/detach of the EBS from one node to the other triggered by draining a swarm node with tasks mounting the volume. 

Debugging this further, I found that in the second/destination node (where the data from the initial node disappeared) the propagation of the EBS mount directory from the plugin container namespace to the base OS or the container that mounted the volume was not happening. As a result, any changes/addition of data in the second node was ending up in a local directory within the second node. Upon migrating back the service to the original node where the mount propagation bug was did not trigger, the original data from the EBS volume was now showing up.

There are a couple of issues that have been fixed recently in the Docker engine in this area. I will test with 17.06.2 engine and see if I can get a repro of this behavior. So far I think this is not a Cloudstor specific issue and may happen with any plugin.

**Comment by ddebroy** [comment]
I did a fair amount of testing with our latest 17.07 edge engine and the underlying engine issue appears to be fixed. Moved around a volume across regions constantly and the data always survived fine. Willl close once the 17.07 release is officially published.

**Comment by FrenchBen** [comment]
@hrajchert Can you confirm that this has been addressed in the latest 17.09 release?
